controller/proxy/proxy.ml

```diff
@@ -0,0 +1,56 @@
+type credentials =
+  { user: string
+  ; password: string
+  }
+
+type t =
+  { credentials: credentials option
+  ; host: string
+  ; port: int
+  }
+
+let validate str =
+  let uri = Uri.of_string str in
+  if Uri.path uri = ""
+    && Uri.query uri = []
+    && Uri.fragment uri = None
+  then
+    match Uri.scheme uri, Uri.host uri, Uri.port uri with
+    | Some "http", Some host, Some port ->
+      Some
+        { credentials =
+          (match Uri.user uri, Uri.password uri with
+          | Some user, Some password -> Some { user; password }
+          | _ -> None)
+        ; host
+        ; port
+        }
+    | _ -> None
+  else
+    None
+
+let to_string ~hide_password t =
+  [ "http://"
+  ; (match t.credentials with
+    | Some credentials ->
+        [ credentials.user
+        ; ":"
+        ; if hide_password then "******" else credentials.password
+        ; "@"
+        ]
+        |> String.concat ""
+    | None -> "")
+  ; t.host
+  ; ":"
+  ; string_of_int t.port
+  ]
+  |> String.concat ""
+
+let from_connected_service services =
+  let open Connman.Service in
+  services
+  |> List.find_map (fun s ->
+    if s.state = Ready || s.state = Online then
```

Comment:
Same remark as in Python version, it might be useful to prefer `Online` services/default gateway.

Reply:
https://github.com/dividat/playos/pull/54/commits/eaf3ac2227e903ea4d71ec72de39d66f297677de

Reply:
Picking the default service in https://github.com/dividat/playos/pull/54/commits/65109fc312bc6f134bd3b2c9046e778b4f2f4c41